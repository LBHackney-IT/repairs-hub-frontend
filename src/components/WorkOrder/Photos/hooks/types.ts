export type Link = {
  presignedUrl: string
  key: string
}
