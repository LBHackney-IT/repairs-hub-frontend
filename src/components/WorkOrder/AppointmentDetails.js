import { useContext } from 'react'
import PropTypes from 'prop-types'
import UserContext from '../UserContext/UserContext'
import Link from 'next/link'
import { dateToStr } from '../../utils/date'
import { STATUS_CANCELLED } from '../../utils/status-codes'
import { priorityCodesRequiringAppointments } from '../../utils/helpers/priorities'
import {
  canSeeAppointmentDetailsInfo,
  canScheduleAppointment,
} from '../../utils/user-permissions'
import { WorkOrder } from '../../models/work-order'

const AppointmentDetails = ({ workOrder, schedulerSessionId }) => {
  const { user } = useContext(UserContext)

  const appointmentDetailsInfoHtml = () => {
    return (
      <span className="govuk-!-font-size-14">
        {dateToStr(new Date(workOrder.appointment.date))},{' '}
        {workOrder.appointment.start}-{workOrder.appointment.end}
      </span>
    )
  }

  const scheduleAppointmentHtml = () => {
    if (workOrder.externalAppointmentManagementUrl) {
      if (schedulerSessionId) {
        return (
          <Link
            href={`${workOrder.externalAppointmentManagementUrl}&sessionId=${schedulerSessionId}`}
          >
            <a className="lbh-link" target="_blank" rel="noopener">
              <strong>Open DRS</strong> to book an appointment
            </a>
          </Link>
        )
      } else {
        console.error('Scheduler Session ID does not exist')
      }
    } else {
      return (
        <Link href={`/work-orders/${workOrder.reference}/appointment/new`}>
          <a className="lbh-link lbh-!-font-weight-bold">
            Schedule an appointment
          </a>
        </Link>
      )
    }
  }

  return (
    (canScheduleAppointment(user) || canSeeAppointmentDetailsInfo(user)) && (
      <div className="appointment-details">
        <span className="govuk-!-font-size-14">Appointment details</span>
        <br></br>
        <div className="lbh-body-s">
          {user && (
            <>
              {canScheduleAppointment(user) &&
                !workOrder.appointment &&
                workOrder.statusAllowsScheduling() &&
                priorityCodesRequiringAppointments.includes(
                  workOrder.priorityCode
                ) &&
                scheduleAppointmentHtml()}
              {canSeeAppointmentDetailsInfo(user) &&
                workOrder.status !== STATUS_CANCELLED &&
                !!workOrder.appointment &&
                appointmentDetailsInfoHtml()}
              {canSeeAppointmentDetailsInfo(user) &&
                !workOrder.appointment &&
                !priorityCodesRequiringAppointments.includes(
                  workOrder.priorityCode
                ) && (
                  <span className="lbh-!-font-weight-bold">Not applicable</span>
                )}
            </>
          )}
        </div>
      </div>
    )
  )
}

AppointmentDetails.propTypes = {
  workOrder: PropTypes.instanceOf(WorkOrder).isRequired,
}

export default AppointmentDetails
