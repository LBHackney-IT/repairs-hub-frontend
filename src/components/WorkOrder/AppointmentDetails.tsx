import { useContext } from 'react'
import UserContext from '../UserContext'
import { STATUS_CANCELLED } from '@/utils/statusCodes'
import {
  canSeeAppointmentDetailsInfo,
  canScheduleAppointment,
} from '@/utils/userPermissions'
import { WorkOrder } from '@/models/workOrder'
import { formatDateTime } from '../../utils/time'
import AppointmentDetailsInfo from './AppointmentDetailsInfo'
import ScheduleAppointment from './ScheduleAppointment/ScheduleAppointment'
import ScheduleInternalAppointmentLink from './ScheduleInternalAppointmentLink'
import { WorkOrderAppointmentDetails } from './types'

interface Props {
  workOrder: WorkOrder
  appointmentDetails: WorkOrderAppointmentDetails
}

const AppointmentDetails = ({ workOrder, appointmentDetails }: Props) => {
  const { user } = useContext(UserContext)

  return (
    <>
      {!!workOrder?.startTime && (
        <div className="lbh-body-xs govuk-!-margin-bottom-2">
          <span>Started at</span>
          <br></br>
          <span>{formatDateTime(workOrder.startTime)}</span>
        </div>
      )}
      {(canScheduleAppointment(user) || canSeeAppointmentDetailsInfo(user)) && (
        <div className="appointment-details" style={{ marginTop: 0 }}>
          {/* <p className="govuk-!-font-size-14">Appointment details</p> */}
          <div className="lbh-body-s govuk-!-margin-0">
            {user && (
              <>
                {canSeeAppointmentDetailsInfo(user) &&
                  appointmentDetails?.appointment &&
                  workOrder.status !== STATUS_CANCELLED && (
                    <AppointmentDetailsInfo
                      appointmentDetails={appointmentDetails}
                    />
                  )}

                {canScheduleAppointment(user) &&
                  workOrder.canBeScheduled() &&
                  (appointmentDetails?.externalAppointmentManagementUrl ? (
                    <ScheduleAppointment
                      hasExistingAppointment={!!appointmentDetails?.appointment}
                      workOrderReference={workOrder.reference}
                      appointmentIsToday={workOrder.appointmentIsToday()}
                      externalAppointmentManagementUrl={
                        appointmentDetails?.externalAppointmentManagementUrl
                      }
                    />
                  ) : (
                    <ScheduleInternalAppointmentLink
                      workOrderReference={workOrder.reference}
                      hasExistingAppointment={appointmentDetails?.appointment}
                      appointmentIsToday={workOrder.appointmentIsToday()}
                    />
                  ))}

                {canSeeAppointmentDetailsInfo(user) &&
                  !appointmentDetails?.appointment &&
                  !workOrder.canBeScheduled() && (
                    <p
                      className="lbh-!-font-weight-bold"
                      style={{ marginTop: 0 }}
                    >
                      Not applicable
                    </p>
                  )}
              </>
            )}
          </div>
        </div>
      )}
    </>
  )
}

export default AppointmentDetails
