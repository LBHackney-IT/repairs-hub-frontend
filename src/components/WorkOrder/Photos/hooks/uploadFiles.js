import axios from 'axios'
import imageCompression from 'browser-image-compression'
import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'

const uploadFiles = async (
  files,
  workOrderReference,
  uploadGroupLabel,
  description,
  setLoadingStatus = null
) => {
  if (setLoadingStatus !== null) {
    setLoadingStatus(`Uploading ${files.length} file(s)`)
  }

  // 1. get presigned urls
  const uploadUrlsResult = await getUploadLinks(
    workOrderReference,
    files.length
  )

  if (!uploadUrlsResult.success) {
    if (setLoadingStatus !== null) setLoadingStatus(null)

    return {
      success: false,
      requestError: uploadUrlsResult.error,
    }
  }

  const presignedUrls = uploadUrlsResult.result.links

  function onProgress(completed, total) {
    setLoadingStatus(`${completed} of ${total} file(s) uploaded`)
  }

  // 2. Upload files to S3
  const uploadFilesToS3Response = await uploadFilesToS3(
    files,
    presignedUrls,
    onProgress
  )

  if (!uploadFilesToS3Response.success) {
    if (setLoadingStatus !== null) setLoadingStatus(null)

    return {
      success: false,
      requestError: uploadFilesToS3Response.error,
    }
  }

  // 3. Complete upload
  const completeUploadResult = await completeUpload(
    workOrderReference,
    presignedUrls.map((x) => x.key),
    uploadGroupLabel,
    description
  )

  if (!completeUploadResult.success) {
    if (setLoadingStatus !== null) setLoadingStatus(null)

    return {
      success: false,
      requestError: completeUploadResult.error,
    }
  }

  return {
    success: true,
  }
}

const completeUpload = async (
  workOrderReference,
  s3Keys,
  uploadGroupLabel,
  description
) => {
  try {
    const result = await frontEndApiRequest({
      method: 'post',
      path: `/api/workOrders/images/completeUpload`,
      requestData: {
        workOrderReference: workOrderReference,
        s3Objects: s3Keys,
        uploadGroupLabel,
        description,
      },
    })

    return { success: true, result }
  } catch (error) {
    return { success: false, error }
  }
}

const uploadFilesToS3 = async (files, links, onProgress) => {
  const promiseList = []
  let completed = 0

  files.forEach((file, i) => {
    promiseList.push(uploadFileToS3(file, links[i]))
  })

  // wrap with callback to track how many have been completed
  const wrappedPromises = promiseList.map((promise) =>
    promise.then((result) => {
      completed += 1
      onProgress(completed, promiseList.length)
      return result
    })
  )

  try {
    const result = await Promise.allSettled(wrappedPromises)

    return { success: true, result }
  } catch (error) {
    return { success: false, error }
  }
}

const uploadFileToS3 = async (file, link) => {
  const compressionOptions = {
    maxSizeMB: 1,
    maxWidthOrHeight: 1920,
    useWebWorker: true,
    maxIteration: 1,
  }

  const compressedFile = await imageCompression(file, compressionOptions)

  return axios.request({
    method: 'PUT',
    url: link.presignedUrl,
    data: compressedFile,
  })
}

const getUploadLinks = async (workOrderReference, numberOfFiles) => {
  try {
    const result = await frontEndApiRequest({
      method: 'get',
      path: `/api/workOrders/images/uploadLink`,
      params: {
        workOrderReference: workOrderReference,
        numberOfFiles: numberOfFiles,
      },
    })

    return { success: true, result }
  } catch (error) {
    return { success: false, error }
  }
}

export default uploadFiles
