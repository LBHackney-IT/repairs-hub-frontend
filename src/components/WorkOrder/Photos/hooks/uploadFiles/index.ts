import getPresignedUrls from './getPresignedUrls'
import uploadFilesToS3 from './uploadFilesToS3'
import completeUpload from './completeUpload'
import { captureException } from '@sentry/nextjs'
import fileUploadStatusLogger from './fileUploadStatusLogger'
import compressFile from './compressFile'
import { getCachedFile, setCachedFile } from './cacheFile'
import ensureAllFilesReadable from './ensureAllFilesReadable'
import useCloudwatchLogger from '@/root/src/utils/cloudwatchLogger'

export class FileUploadError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'FileUploadError'
  }
}

const uploadFiles = async (
  files: File[],
  workOrderReference: string,
  description: string,
  uploadGroupLabel: string,
  setUploadStatus: (status: string) => void,
  cwLogger: ReturnType<typeof useCloudwatchLogger>
): Promise<{
  success: boolean
  requestError?: string
}> => {
  try {
    const statusLogger = fileUploadStatusLogger(files.length, setUploadStatus)

    // 1. Get cached files to upload or compress and cache them
    const filesToUpload: File[] = []
    const compressionErrors: Error[] = []

    for (const file of files) {
      try {
        let fileToUpload = await getCachedFile(file)

        if (fileToUpload == null) {
          fileToUpload = await compressFile(file)
          await setCachedFile(fileToUpload)
        }

        filesToUpload.push(fileToUpload)
      } catch (error) {
        filesToUpload.push(file)
        compressionErrors.push(error as Error)
      } finally {
        statusLogger('Compress')
      }
    }

    // 2. Get presigned urls
    const uploadUrlsResult = await getPresignedUrls(
      workOrderReference,
      filesToUpload.length
    )
    if (!uploadUrlsResult.success)
      throw new FileUploadError(uploadUrlsResult.error as string)

    const presignedUrls = uploadUrlsResult.result.links

    // 3. Upload files to S3
    await ensureAllFilesReadable(filesToUpload)
    const uploadFilesToS3Response = await uploadFilesToS3(
      filesToUpload,
      presignedUrls,
      () => statusLogger('Upload')
    )
    if (!uploadFilesToS3Response.success)
      throw new FileUploadError(
        (uploadFilesToS3Response.error as string) +
          (compressionErrors.length &&
            `| Compression error: ${compressionErrors?.[0]?.message}`)
      )

    // 4. Complete upload
    const completeUploadResult = await completeUpload(
      workOrderReference,
      presignedUrls.map((x) => x.key),
      description,
      uploadGroupLabel
    )
    if (!completeUploadResult.success)
      throw new FileUploadError(completeUploadResult.error as string)
  } catch (error) {
    if (error instanceof FileUploadError) {
      cwLogger.error(error.message)
      captureException('Failed to upload photos', {
        tags: {
          section: 'File upload',
        },
        extra: {
          workOrderReference,
          files: JSON.stringify(
            files.map((file) => ({
              name: file.name,
              size: file.size,
              type: file.type,
            }))
          ),
          message: error.message,
        },
      })
    }
    return {
      success: false,
      requestError: error.message || 'An error occurred while uploading files',
    }
  }

  return {
    success: true,
  }
}

export default uploadFiles
