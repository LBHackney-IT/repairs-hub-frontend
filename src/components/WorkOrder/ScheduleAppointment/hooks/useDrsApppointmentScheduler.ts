import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'
import { buildDataFromScheduleAppointment } from '@/root/src/utils/hact/jobStatusUpdate/notesForm'
import { useContext, useEffect, useRef, useState } from 'react'
import UserContext from '../../../UserContext'
import { getOrCreateSchedulerSessionId } from '../../../../utils/frontEndApiClient/users/schedulerSession'

export const useDrsAppointmentScheduler = (workOrderReference: string) => {
  const [schedulerSessionId, setSchedulerSessionId] = useState<string | null>(
    null
  )
  const [isLoading, setIsLoading] = useState<boolean>(true)
  const [error, setError] = useState<string | null>(null)

  const { user } = useContext(UserContext)
  const userOpenedWBM = useRef(false)

  const fetchSchedulerSessionId = async () => {
    setIsLoading(true)

    try {
      const sessionId = await getOrCreateSchedulerSessionId()
      setSchedulerSessionId(sessionId)
    } catch (err) {
      console.error(err)
      setError(
        `Something went wrong fetching DRS Session: ${
          err?.message || 'Unknown error'
        }`
      )
    } finally {
      setIsLoading(false)
    }
  }

  const handleExternalLinkOpen = async () => {
    userOpenedWBM.current = true

    const jobStatusUpdate = buildDataFromScheduleAppointment(
      workOrderReference.toString(),
      `${user.name} opened the DRS Web Booking Manager`
    )

    await frontEndApiRequest({
      method: 'post',
      path: '/api/jobStatusUpdate',
      requestData: jobStatusUpdate,
    })

    setTimeout(() => fetchSchedulerSessionId(), 500)
  }

  useEffect(() => {
    const handleWindowFocus = () => {
      if (userOpenedWBM.current) {
        window.location.reload()
      }
    }

    window.addEventListener('focus', handleWindowFocus)

    return () => {
      window.removeEventListener('focus', handleWindowFocus)
    }
  }, [])

  useEffect(() => {
    fetchSchedulerSessionId()
  }, [])

  return {
    schedulerSessionId,
    isLoading,
    error,
    handleExternalLinkOpen,
  }
}
