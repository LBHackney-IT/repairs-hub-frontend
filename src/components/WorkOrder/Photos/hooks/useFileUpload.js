import { useState } from 'react'
import uploadFiles from './uploadFiles'
import validateFileUpload from './validateFileUpload'

const useFileUpload = (workOrderReference, onSuccess) => {
  const [loadingStatus, setLoadingStatus] = useState(null)
  const [validationError, setValidationError] = useState(null)
  const [requestError, setRequestError] = useState(null)
  const [uploadSuccess, setUploadSuccess] = useState(null)

  const [files, setFiles] = useState([])

  const handleSubmit = async (e) => {
    e.preventDefault()

    if (loadingStatus) return

    // reset
    setUploadSuccess(null)
    setRequestError(null)
    setValidationError(null)

    const validationResult = validateFileUpload(files)
    setValidationError(validationResult)
    if (validationResult !== null) return

    const description = e.target.description.value

    const uploadResult = await uploadFiles(
      files,
      workOrderReference,
      '',
      'Uploaded directly to work order',
      description,
      setLoadingStatus
    )

    setLoadingStatus(null)

    if (!uploadResult.success) {
      setRequestError(uploadResult.requestError.message)
      return
    }

    setUploadSuccess(
      ` ${files.length} ${
        files.length === 1 ? 'photo has' : 'photos have'
      } been added to the workOrder`
    )
    setFiles([])
    onSuccess()
  }

  return {
    files,
    handleSubmit,
    uploadSuccess,
    setFiles,
    validationError,
    requestError,
    loadingStatus,
  }
}

export default useFileUpload
