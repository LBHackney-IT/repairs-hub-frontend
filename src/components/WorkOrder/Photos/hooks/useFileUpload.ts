import { useState } from 'react'
import uploadFiles from './uploadFiles'
import validateFileUpload from './validateFileUpload'

const useFileUpload = (workOrderReference: string, onSuccess: () => void) => {
  const [loadingStatus, setLoadingStatus] = useState(null)
  const [validationError, setValidationError] = useState(null)
  const [requestError, setRequestError] = useState(null)
  const [uploadSuccess, setUploadSuccess] = useState(null)

  const [files, setFiles] = useState<File[]>([])

  const handleSubmit = async (e) => {
    e.preventDefault()

    if (loadingStatus) return

    // reset
    setUploadSuccess(null)
    setRequestError(null)
    setValidationError(null)

    const validationResult = validateFileUpload(files)
    setValidationError(validationResult)
    if (validationResult !== null) return

    const description = e.target.description.value

    const uploadResult = await uploadFiles(
      files,
      workOrderReference,
      description,
      'Uploaded directly to work order',
      setLoadingStatus
    )

    setLoadingStatus(null)

    if (!uploadResult.success) {
      setRequestError(uploadResult.requestError)
      return
    }

    setUploadSuccess(
      ` ${files.length} ${
        files.length === 1 ? 'photo has' : 'photos have'
      } been added to the work order`
    )
    setFiles([])
    onSuccess()
  }

  return {
    files,
    handleSubmit,
    uploadSuccess,
    setFiles,
    validationError,
    requestError,
    loadingStatus,
  }
}

export default useFileUpload
