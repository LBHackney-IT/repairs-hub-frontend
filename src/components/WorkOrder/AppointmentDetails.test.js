jest.mock('./ScheduleAppointment/hooks/useDrsApppointmentScheduler', () => ({
  useDrsAppointmentScheduler: jest.fn(),
}))
import { useDrsAppointmentScheduler } from './ScheduleAppointment/hooks/useDrsApppointmentScheduler'

import { render } from '@testing-library/react'
import UserContext from '../UserContext'
import { agent } from 'factories/agent'
import AppointmentDetails from './AppointmentDetails'
import { WorkOrder } from '@/models/workOrder'

const workOrderData = {
  reference: 10000012,
  dateRaised: '2021-01-18T15:28:57.17811',
  lastUpdated: null,
  priority: 'A priority',
  property: '16 Pitcairn House  St Thomass Square',
  owner: 'HH General Building Repair - H01',
  description: 'This is a repair description',
  propertyReference: '00012345',
  status: 'In Progress',
  priorityCode: 0,
  raisedBy: 'Dummy Agent',
  target: '2021-01-23T18:30:00.00000',
  tradeCode: 'PL',
  tradeDescription: 'PLUMBING - PL',
  callerName: 'Jill Smith',
  callerNumber: '07700 900999',
  contractorReference: 'H01',
  appointment: null,
}

const schedulerSessionId = 'SCHEDULER_SESSION_ID'

describe('AppointmentDetails component', () => {
  const appointment = {
    date: '2021-03-19',
    description: 'PM Slot',
    end: '18:00',
    start: '12:00',
  }

  beforeEach(() => {
    jest.clearAllMocks()

    useDrsAppointmentScheduler.mockReturnValue({
      schedulerSessionId: schedulerSessionId,
      isLoading: false,
      error: null,
      handleExternalLinkOpen: jest.fn(),
    })
  })

  describe('DRS work order', () => {
    let drsWorkOrder = {
      ...workOrderData,
      externalAppointmentManagementUrl:
        'https://scheduler.example.hackney.gov.uk?bookingId=1',
    }

    describe('with no appointment', () => {
      describe('when the work order can be scheduled', () => {
        it('shows a link to schedule an appointment with DRS Web Booking Manager', () => {
          let workOrder = new WorkOrder(drsWorkOrder)

          workOrder.canBeScheduled = jest.fn(() => true)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })

      describe('when the work order cannot be scheduled', () => {
        it('does not show a schedule link', () => {
          let workOrder = new WorkOrder(drsWorkOrder)

          workOrder.canBeScheduled = jest.fn(() => false)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })
    })

    describe('with an appointment', () => {
      describe('when the work order can be scheduled', () => {
        it('shows a link to reschedule an appointment with DRS Web Booking Manager', () => {
          let workOrder = new WorkOrder({ ...drsWorkOrder, appointment })

          workOrder.canBeScheduled = jest.fn(() => true)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })

      describe('when the work order cannot be scheduled', () => {
        it('does not show a reschedule link but shows the existing appointment', () => {
          let workOrder = new WorkOrder({ ...drsWorkOrder, appointment })

          workOrder.canBeScheduled = jest.fn(() => false)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })
    })
  })

  describe('Work order (non DRS)', () => {
    let nonDRSWorkOrder = {
      ...workOrderData,
      externalAppointmentManagementUrl: null,
    }

    describe('with no appointment', () => {
      describe('when the work order can be scheduled', () => {
        it('shows a link to schedule an appointment', () => {
          let workOrder = new WorkOrder(nonDRSWorkOrder)

          workOrder.canBeScheduled = jest.fn(() => true)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })

      describe('when the work order cannot be scheduled', () => {
        it('does not show a schedule link', () => {
          let workOrder = new WorkOrder(nonDRSWorkOrder)

          workOrder.canBeScheduled = jest.fn(() => false)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })
    })

    describe('with an appointment', () => {
      describe('when the work order can be scheduled', () => {
        it('shows a link to reschedule an appointment', () => {
          let workOrder = new WorkOrder({ ...nonDRSWorkOrder, appointment })

          workOrder.canBeScheduled = jest.fn(() => true)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })

      describe('when the work order cannot be scheduled', () => {
        it('does not show a reschedule link but shows the existing appointment', () => {
          let workOrder = new WorkOrder({ ...nonDRSWorkOrder, appointment })

          workOrder.canBeScheduled = jest.fn(() => false)

          const { asFragment } = render(
            <UserContext.Provider value={{ user: agent }}>
              <AppointmentDetails workOrder={workOrder} />
            </UserContext.Provider>
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })
    })
  })
})
