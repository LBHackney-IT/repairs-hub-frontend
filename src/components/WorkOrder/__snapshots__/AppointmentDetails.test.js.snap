// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentDetails component DRS work order with an appointment when the work order can be scheduled shows a link to reschedule an appointment with DRS Web Booking Manager 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <a
        class="lbh-link"
        href="/work-orders/10000012/appointment/new"
      >
        Schedule appointment (non DRS)
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppointmentDetails component DRS work order with an appointment when the work order cannot be scheduled does not show a reschedule link but shows the existing appointment 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <p
        class="lbh-!-font-weight-bold"
        style="margin-top: 0px;"
      >
        Not applicable
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppointmentDetails component DRS work order with no appointment when the work order can be scheduled shows a link to schedule an appointment with DRS Web Booking Manager 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <a
        class="lbh-link"
        href="/work-orders/10000012/appointment/new"
      >
        Schedule appointment (non DRS)
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppointmentDetails component DRS work order with no appointment when the work order cannot be scheduled does not show a schedule link 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <p
        class="lbh-!-font-weight-bold"
        style="margin-top: 0px;"
      >
        Not applicable
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppointmentDetails component Work order (non DRS) with an appointment when the work order can be scheduled shows a link to reschedule an appointment 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <a
        class="lbh-link"
        href="/work-orders/10000012/appointment/new"
      >
        Schedule appointment (non DRS)
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppointmentDetails component Work order (non DRS) with an appointment when the work order cannot be scheduled does not show a reschedule link but shows the existing appointment 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <p
        class="lbh-!-font-weight-bold"
        style="margin-top: 0px;"
      >
        Not applicable
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppointmentDetails component Work order (non DRS) with no appointment when the work order can be scheduled shows a link to schedule an appointment 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <a
        class="lbh-link"
        href="/work-orders/10000012/appointment/new"
      >
        Schedule appointment (non DRS)
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppointmentDetails component Work order (non DRS) with no appointment when the work order cannot be scheduled does not show a schedule link 1`] = `
<DocumentFragment>
  <div
    class="appointment-details"
    style="margin-top: 0px;"
  >
    <p
      class="govuk-!-font-size-14"
    >
      Appointment details
    </p>
    <div
      class="lbh-body-s govuk-!-margin-0"
    >
      <p
        class="lbh-!-font-weight-bold"
        style="margin-top: 0px;"
      >
        Not applicable
      </p>
    </div>
  </div>
</DocumentFragment>
`;
